pylzy/lzy/env/explorer/classify.py

```diff
@@ -0,0 +1,261 @@
+from __future__ import annotations
+
+import json
+import os
+from functools import lru_cache
+from typing import FrozenSet, Set, Dict, cast
+from types import ModuleType
+
+from importlib.machinery import ExtensionFileLoader
+from importlib_metadata import Distribution
+
+from lzy.utils.pypi import check_package_version_exists
+
+from .search import ModulesFrozenSet, ModulesSet
+from .packages import (
+    LocalPackage,
+    LocalDistribution,
+    PypiDistribution,
+    BrokenModules,
+    BasePackage,
+)
+from .utils import (
+    get_files_to_distributions,
+    get_stdlib_module_names,
+    get_builtin_module_names,
+    check_url_is_local_file,
+)
+
+
+class ModuleClassifier:
+    def __init__(self, pypi_index_url: str):
+        self.pypi_index_url = pypi_index_url
+
+        self.distributions: Set[Distribution]
+        self.modules_without_distribution: ModulesSet
+        self.binary_distributions: Set[Distribution]
+
+        self.stdlib_module_names = get_stdlib_module_names()
+        self.builtin_module_names = get_builtin_module_names()
+        self.files_to_distributions = get_files_to_distributions()
+
+    def __call__(self, modules: ModulesFrozenSet) -> FrozenSet[BasePackage]:
+        self.distributions = set()
+        self.binary_distributions = set()
+        self.modules_without_distribution = set()
+
+        self._classify_modules(modules)
+        packages = self._classify_distributions()
+        packages |= self._classify_modules_without_distributions()
+
+        return frozenset(packages)
+
+    def _classify_modules(self, modules: ModulesFrozenSet) -> None:
+        """
+        Here we are dividing modules into two piles:
+        those which are part of some distribution and those which are not.
+        Also here we are noting distributions with binary modules.
+        """
+        for module in modules:
+            module_name = module.__name__
+            top_level: str = module_name.split('.')[0]
+            filename = getattr(module, '__file__', None)
+
+            # Modules without __file__ doesn't represent specific file at the disk
+            # so we a generally doesn't interested about it.
+            # It can be namespace modules or strange virtual modules as the `six.moves.*`.
+            if not filename:
```

Comment:
Maybe add warning here?
Are this modules always part of other module?

Reply:
And maybe add trace logs to debug module classification and searching?

Reply:
1) No, no warnings are needed. It's a normal situation when a module doesn't have a `__file__` attribute, for example - namespace modules like `google`. 
2) I'm planning to add logging to Explorer, but I still thinking about logging here.